src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -6322,12 +6561,190 @@ private IOperation VisitNoneOperationExpression(IOperation operation)
+
+            _currentInterpolatedStringHandlerId = previousHandlerId;
+            return new FlowCaptureReferenceOperation(handlerCaptureId, operation.Syntax, operation.Type, operation.GetConstantValue());
+
+            static ImmutableArray<IInterpolatedStringAppendOperation> collectAppendCalls(IInterpolatedStringHandlerCreationOperation creation)
+            {
+                var appendCalls = ArrayBuilder<IInterpolatedStringAppendOperation>.GetInstance();
+                if (creation.Content is IInterpolatedStringOperation interpolatedString)
+                {
+                    // Simple case
+                    appendStringCalls(interpolatedString, appendCalls);
+                    return appendCalls.ToImmutableAndFree();
+                }
+
+                var stack = ArrayBuilder<IInterpolatedStringAdditionOperation>.GetInstance();
+                pushLeftNodes((IInterpolatedStringAdditionOperation)creation.Content, stack);
+
+                while (stack.TryPop(out var currentAddition))
+                {
+                    switch (currentAddition.Left)
+                    {
+                        case IInterpolatedStringOperation interpolatedString1:
+                            appendStringCalls(interpolatedString1, appendCalls);
+                            break;
+                        case IInterpolatedStringAdditionOperation:
+                            break;
+                        default:
+                            throw ExceptionUtilities.UnexpectedValue(currentAddition.Left.Kind);
+                    }
+
+                    switch (currentAddition.Right)
+                    {
+                        case IInterpolatedStringOperation interpolatedString1:
+                            appendStringCalls(interpolatedString1, appendCalls);
+                            break;
+                        case IInterpolatedStringAdditionOperation additionOperation:
+                            pushLeftNodes(additionOperation, stack);
+                            break;
+                        default:
+                            throw ExceptionUtilities.UnexpectedValue(currentAddition.Left.Kind);
+                    }
+                }
+
+                stack.Free();
+                return appendCalls.ToImmutableAndFree();
+
+                static void appendStringCalls(IInterpolatedStringOperation interpolatedString, ArrayBuilder<IInterpolatedStringAppendOperation> appendCalls)
+                {
+                    foreach (var part in interpolatedString.Parts)
+                    {
+                        appendCalls.Add((IInterpolatedStringAppendOperation)part);
+                    }
+                }
+
+                static void pushLeftNodes(IInterpolatedStringAdditionOperation addition, ArrayBuilder<IInterpolatedStringAdditionOperation> stack)
+                {
+                    IInterpolatedStringAdditionOperation? current = addition;
+                    while (current != null)
```

Comment:
>while (current != null) [](http://example.com/codeflow?start=20&length=23)

We know that we will do at least one iteration. In cases like that a post condition should be used, i.e. ```do while```